### PR TITLE
[GLUTEN-7837][VL] Spark driver should not initialize cache if not in local mode

### PR DESCRIPTION
**Pull request by leoluan2009**
## What changes were proposed in this pull request?

Fixes: https://github.com/apache/incubator-gluten/issues/7837

Spark driver should not initialize cache if not in local mode

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by leoluan2009** [comment]
@zhztheplayer  can you help review?

**Comment by zhztheplayer** [comment]
@leoluan2009 Does this solve the cache issue mentioned in https://github.com/apache/incubator-gluten/issues/7837?

**Comment by leoluan2009** [comment]
> @leoluan2009 Does this solve the cache issue mentioned in #7837?

Yes, if VeloxBackend is running in driver, it will not init cache.

**Review by yikf — APPROVED** [review]

File: cpp/velox/compute/VeloxBackend.cc [hunk 1]
```diff
@@ -195,7 +195,7 @@ void VeloxBackend::initJolFilesystem() {
 }
 
 void VeloxBackend::initCache() {
-  if (backendConf_->get<bool>(kVeloxCacheEnabled, false)) {
+  if (backendConf_->get<bool>(kVeloxCacheEnabled, false) && !isDriver_) {
```
Comment (zhztheplayer) [inline cpp/velox/compute/VeloxBackend.cc:198]: Could just change from the caller code, something like:

```
if (!isDriver_) {
  initCache();
}
...
```

Would be enough and more readable.

**Review by zhztheplayer — COMMENTED** [review]
How does the change work with Spark local mode? I mean we may need to turn the cache on in local mode if user doesn't explicitly disable it.

File: cpp/velox/compute/VeloxBackend.cc (hunk 1 above)
Comment (leoluan2009) [reply cpp/velox/compute/VeloxBackend.cc]: Done

**Comment by zhztheplayer** [comment]
Hi @leoluan2009 , as `spark.gluten.sql.columnar.backend.velox.cacheEnabled` is now a static SQL conf, can we just modify the config map passed to native to set it to false? It's a workaround and we can remove once a optimal design is out.

**Comment by leoluan2009** [comment]
> Hi @leoluan2009 , as `spark.gluten.sql.columnar.backend.velox.cacheEnabled` is now a static SQL conf, can we just modify the config map passed to native to set it to false from non-local driver side? It's a workaround and we can remove once a optimal design is out.

it's ok and I will update the pr

**Comment by leoluan2009** [comment]
@zhztheplayer  I have updated this PR, please help to review, thanks!

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxListenerApi.scala [hunk 2]
```diff
@@ -157,7 +155,10 @@ class VeloxListenerApi extends ListenerApi with Logging {
     }
 
     // Initial native backend with configurations.
-    val parsed = GlutenConfigUtil.parseConfig(conf.getAll.toMap)
+    var parsed = GlutenConfigUtil.parseConfig(conf.getAll.toMap)
+    if (isDriver && !inLocalMode(conf)) {
+      parsed += (GlutenConfig.COLUMNAR_VELOX_CACHE_ENABLED.key -> "false")
```
Comment (zhztheplayer) [inline backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxListenerApi.scala:160]: Would you like to add some comments in code? Or just leave a link to the relevant issue/PR. Thanks.

**Review by zhztheplayer — APPROVED** [review]
:+1

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxListenerApi.scala [hunk 3]
```diff
@@ -26,7 +26,6 @@ import org.apache.gluten.init.NativeBackendInitializer
 import org.apache.gluten.jni.{JniLibLoader, JniWorkspace}
 import org.apache.gluten.udf.UdfJniWrapper
 import org.apache.gluten.utils._
-
```
Comment (jackylee-ch) [inline backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxListenerApi.scala:29]: This line should not be removed.

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxListenerApi.scala [hunk 4]
```diff
@@ -36,7 +35,6 @@ import org.apache.spark.sql.execution.datasources.velox.{VeloxOrcWriterInjects,
 import org.apache.spark.sql.expression.UDFResolver
 import org.apache.spark.sql.internal.{GlutenConfigUtil, StaticSQLConf}
 import org.apache.spark.util.{SparkDirectoryUtil, SparkResourceUtil}
-
```
Comment (jackylee-ch) [inline backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxListenerApi.scala:39]: ditto

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxListenerApi.scala (hunk 3 above)
Comment (leoluan2009) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxListenerApi.scala]: Done, thanks!

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxListenerApi.scala (hunk 4 above)
Comment (leoluan2009) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxListenerApi.scala]: Done, thanks!

File: backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxListenerApi.scala (hunk 2 above)
Comment (leoluan2009) [reply backends-velox/src/main/scala/org/apache/gluten/backendsapi/velox/VeloxListenerApi.scala:163]: Done, thanks

**Comment by leoluan2009** [comment]
@zhztheplayer  I resolved all comments

**Comment by FelixYBW** [comment]
@zhztheplayer @leoluan2009
can you go through all the initializations there? Is there any other things we shouldn't run on driver? Is the local cache only exception? I'd expect the initialization should be much different.

If we have more than one, we should considered to seperate the initialization routine for driver and worker.

**Comment by zhztheplayer** [comment]
> can you go through all the initializations there? Is there any other things we shouldn't run on driver? Is the local cache only exception? I'd expect the initialization should be much different.

Not much difference so far and it's also about our C++ API's encapsulation. We can make trade-offs once it's really needed.

**Comment by zhztheplayer** [comment]
> If we have more than one, we should considered to seperate the initialization routine for driver and worker.

In future there will be different ways to implement differentiated initialization, but exposing Spark arch conceptions "driver" /  "executor" / "task" to C++ code could be the last ones to consider. We can use configurations, or compose finer-grained JNI initialization APIs, etc.

**Comment by FelixYBW** [comment]
> In future there will be different ways to implement differentiated initialization, but exposing Spark arch conceptions "driver" / "executor" / "task" to C++ code could be the last ones to consider. We can use configurations, or compose finer-grained JNI initialization APIs, etc.

Let's re-evaluate the mythology once we have another exception.

**Pull request merged**